spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -154,6 +156,22 @@ else if (this.listenerConsumer.assignedPartitions != null) {
 	}
 
 	@Override
+	public Map<String, Map<MetricName, Metric>> metrics() {
+		ListenerConsumer listenerConsumer = this.listenerConsumer;
+		Map<String, Map<MetricName, Metric>> metrics = new HashMap<>();
+		if (listenerConsumer != null) {
+			for (Map.Entry<MetricName, ? extends Metric> entry : listenerConsumer.consumer.metrics().entrySet()) {
+				MetricName metricName = entry.getKey();
+				Metric metric = entry.getValue();
+				String client = metricName.tags().get("client-id");
+				Map<MetricName, Metric> clientMetrics = metrics.computeIfAbsent(client, k -> new HashMap<>());
```

Comment:
If you want to have it backported, the code has to be Java 7 compatible.

And please, consider do not squash commits in the future. That way it easier to track the changes in the PR and we'd have a notification about a new commit. Right now I've just decided to take a look into the code one more time and realized there is a new fix! 
We squash during merge into the `master` anyway. Thanks.

On the other hand, I think the way you extract the `clientId` is fine. Here is the code from the `KafkaConsumer`:
```
String clientId = config.getString(ConsumerConfig.CLIENT_ID_CONFIG);
if (clientId.length() <= 0)
                clientId = "consumer-" + CONSUMER_CLIENT_ID_SEQUENCE.getAndIncrement();
this.clientId = clientId;
Map<String, String> metricsTags = Collections.singletonMap("client-id", clientId);
```
That's sad that there is no `getClientId()`, but we really can use that tag from metrics.